docs/develop/python/message-passing.mdx

```diff
@@ -431,113 +474,121 @@ class MyWorkflow:
+    This happens if the Workflow finished while the Update handler execution was in progress, for example because
 
-    This will happen if the Workflow finished while the Update handler execution was in progress, for example because
     - The Workflow was canceled or failed.
-    - The Workflow completed normally or continued-as-new and the Workflow author did not [wait for handlers to be finished](/encyclopedia/workflow-message-passing#finishing-message-handlers).
-
-### Problems when sending a Query
-
-1. **There is no Workflow Worker polling the Task Queue**
 
-    You'll get an [`temporalio.service.RPCError`](https://python.temporal.io/temporalio.service.RPCError.html) on which the `status` attribute is [`RPCStatusCode`](https://python.temporal.io/temporalio.service.RPCStatusCode.html) `FAILED_PRECONDITION`.
+    - The Workflow completed normally or continued-as-new and the Workflow author did not [wait for handlers to be finished](/encyclopedia/workflow-message-passing#finishing-message-handlers).
 
-1. **Query failed.**
+### Problems when sending a Query {#query-problems}
 
-    You'll get [`temporalio.client.WorkflowQueryFailedError`](https://python.temporal.io/temporalio.client.WorkflowQueryFailedError.html).
-    Any exception in a Query handler will cause this.
-    Note that this differs from Signal and Update, for which exceptions can cause Workflow Task Failure.
+- **There is no Workflow Worker polling the Task Queue**:
+    You'll receive a [`temporalio.service.RPCError`](https://python.temporal.io/temporalio.service.RPCError.html) exception on which the `status` attribute is [`RPCStatusCode`](https://python.temporal.io/temporalio.service.RPCStatusCode.html) `FAILED_PRECONDITION`.
 
-1. **The handler caused the Workflow Task to fail.**
+- **Query failed**:
+  You'll receive a [`temporalio.client.WorkflowQueryFailedError`](https://python.temporal.io/temporalio.client.WorkflowQueryFailedError.html) exception if something goes wrong during a Query.
+  Any exception in a Query handler will trigger this error.
+  This differs from Signal and Update requests, where exceptions can lead to Workflow Task Failure instead.
 
+- **The handler caused the Workflow Task to fail.**
     This would happen, for example, if the Query handler blocks the thread for too long without yielding.
 
+## Dynamic handlers {#dynamic-handler}
 
-## Dynamic Handlers {#dynamic-handler}
-
-Temporal supports Dynamic Workflows, Activities, Signals, Updates, and Queries.
-These are unnamed handlers that are invoked if no other statically defined handler with the given name exists.
-
-Dynamic Handlers provide flexibility to handle cases where the names of Workflows, Activities, Signals, or Queries aren't known at run time.
+Dynamic Workflows, Activities, Signals, Updates, and Queries are a kind of unnamed handler.
+They're looked up by name and dynamically invoked at runtime.
```

Comment:
You give a good explanation of what dynamic handlers are below. I think that explanation needs to be hoisted up here.

IMO it requires considerable sophistication to know how "dynamically invoked at runtime" differs from simply "invoked at runtime", and all handlers by definition are invoked at runtime, so I vote we drop that phrase entirely (here and below).

Reply:
hoisted